### PR TITLE
Feature/pod value mount

### PR DESCRIPTION
**Pull request by ycliuhw**
### Checklist

 - [x] Checked if it requires a [pylibjuju](https://github.com/juju/python-libjuju) change?
 - [ ] Added [integration tests](https://github.com/juju/juju/tree/develop/tests) for the PR?
 - [ ] Added or updated [doc.go](https://discourse.jujucharms.com/t/readme-in-packages/451) related to packages changed?
 - [x] Do comments answer the question of why design decisions were made?

----

## Description of change

Introduce `config` options in k8sspec for environment value mount using `Env` and `EnvFrom`.

## QA steps

deploy a CaaS application with below k8s spec format;

```yaml
containers:
  - name: gitlab
    image: gitlab/latest
    config:
      attr: foo=bar; name["fred"]="blogs";
      foo: bar
      brackets: '["hello", "world"]'
      restricted: 'yes'
      switch: on
      bar: true
      special: p@ssword's
      foo: bar
      float: 111.11111111
      int: 111
      MY_NODE_NAME:
        field:
          path: spec.nodeName
          api-version: v1
      my-resource-limit:
        resource:
          container-name: container1
          resource: requests.cpu
          divisor: 1m
      thing:
        secret:
          name: foo
          key: bar
      a-secret:
        secret:
          name: secret1
          optional: true
      another-secret:
        secret:
          name: secret2
      thing1:
        config-map:
          name: foo
          key: bar
      a-configmap:
        config-map:
          name: configmap1
          optional: true
      another-configmap:
        config-map:
          name: configmap2
```

```console
$ cat reactive/spec_template.yaml
version: 2
omitServiceFrontend: true
containers:
  - name: %(name)s
    imageDetails:
      imagePath: %(docker_image_path)s
      username: %(docker_image_username)s
      password: %(docker_image_password)s
    ports:
      - containerPort: %(port)s
        protocol: TCP
    config:
      MYSQL_ROOT_PASSWORD: %(root_password)s
      MYSQL_USER: %(user)s
      MYSQL_PASSWORD: %(password)s
      MYSQL_DATABASE: %(database)s
      MY_NODE_NAME:
        field:
          path: spec.nodeName
          api-version: v1
      build-robot-secret:
        secret:
          name: build-robot-secret
      ...

$ mkubectl get pod/mariadb-k8s-0 -n t1 -o jsonpath="{..env}"
[map[name:MYSQL_DATABASE value:database] map[name:MYSQL_PASSWORD value:password] map[name:MYSQL_ROOT_PASSWORD value:root] map[name:MYSQL_USER value:admin] map[name:MY_NODE_NAME valueFrom:map[fieldRef:map[apiVersion:v1 fieldPath:spec.nodeName]]] map[name:build-robot-secret valueFrom:map[secretKeyRef:map[key:config.yaml name:build-robot-secret]]]]

$ juju run --unit mariadb-k8s/2 --timeout=30s "env" --operator=false -m k1:t1
MYSQL_ROOT_PASSWORD=root
MYSQL_USER=admin
MYSQL_PASSWORD=password
MYSQL_DATABASE=database
MY_NODE_NAME=ubuntu
username=admin  # from `secret/build-robot-secret`
password=1f2d1e2e67df  # from `secret/build-robot-secret`
...

```

## Documentation changes

Yes

## Bug reference

https://bugs.launchpad.net/juju/+bug/1858515

### CODE REVIEW

**Comment by ycliuhw** [comment]
[k8s valueFrom in pod config](https://trello.com/c/0zgyiGhv/997-k8s-valuefrom-in-pod-config)

File: caas/kubernetes/provider/specs/types.go [hunk 1]
```diff
@@ -72,6 +72,9 @@ type K8sContainerSpec struct {
 	LivenessProbe   *core.Probe           `json:"livenessProbe,omitempty" yaml:"livenessProbe,omitempty"`
 	ReadinessProbe  *core.Probe           `json:"readinessProbe,omitempty" yaml:"readinessProbe,omitempty"`
 	SecurityContext *core.SecurityContext `json:"securityContext,omitempty" yaml:"securityContext,omitempty"`
+
+	EnvFrom []core.EnvFromSource `json:"envFrom,omitempty" yaml:"envFrom,omitempty"`
+	Env     []core.EnvVar        `json:"env,omitempty" yaml:"env,omitempty"`
```
Comment (wallyworld) [inline caas/kubernetes/provider/specs/types.go]: We need a slightly different abstraction here.
We already support the "config" keyword for straight pod env key values.

ie
```
    config:
      attr: foo=bar; name["fred"]="blogs";
      foo: bar
```

We need to look to extend that to add in the extra capability, eg

```
    config:
      foo: bar
      my_node-name:
        field:
          name: spec.nodeName
      thing:
         secret:
           name: foo
           key: bar
      thing2:
         secret:
           name: foo
      thing3:
         configMap:
           name: foo
      thing4:
         configMap:
           name: foo
           value: bar
```

So "config" attributes are extended to accept either a string as we have now for simple scalar values, or a map which is used to define the 3 new cases.
There are 3 new keywords: field, configMap, secret.
If only the name is specified for configMap or secret, that maps to EnvFrom where all fields are used. Otherwise we just use the single named field.

File: caas/kubernetes/provider/k8s_test.go [hunk 2]
```diff
@@ -279,10 +490,10 @@ func getBasicPodspec() *specs.PodSpec {
 		WorkingDir:   "/path/to/here",
 		Config: map[string]interface{}{
 			"foo":        "bar",
-			"restricted": "'yes'",
+			"restricted": "yes",
 			"bar":        true,
-			"switch":     "on",
-			"brackets":   `'["hello", "world"]'`,
+			"switch":     true,
```
Comment (wallyworld) [inline caas/kubernetes/provider/k8s_test.go]: why this change?

File: caas/specs/types.go [hunk 3]
```diff
@@ -52,6 +52,9 @@ type ImageDetails struct {
 // PullPolicy describes a policy for if/when to pull a container image.
 type PullPolicy string
 
+// ContainerConfig describes a config map for a container.
```
Comment (wallyworld) [inline caas/specs/types.go]: ContainerConfig describes the config used for setting up a pod container's environment variables.

File: caas/kubernetes/provider/specs/v2_test.go [hunk 4]
```diff
@@ -73,6 +73,11 @@ containers:
       restricted: 'yes'
       switch: on
       special: p@ssword's
+      my-resource-limit:
```
Comment (wallyworld) [inline caas/kubernetes/provider/specs/v2_test.go]: resource limits are handled via juju constraints - we don't want them in here

File: caas/kubernetes/provider/specs/types.go [hunk 5]
```diff
@@ -100,26 +98,6 @@ func (ps PodSpec) IsEmpty() bool {
 		ps.DNSPolicy == ""
 }
 
-var boolValues = set.NewStrings(
```
Comment (wallyworld) [inline caas/kubernetes/provider/specs/types.go]: we still need the bool value mapping
yes -> true
on -> true 
etc
and if the user enters 'yes', then that's treated as a string

File: caas/kubernetes/provider/specs/v2_test.go (hunk 4 above)
Comment (wallyworld) [reply caas/kubernetes/provider/specs/v2_test.go]: ah, sorry, i misunderstood

File: caas/kubernetes/provider/specs/types.go (hunk 5 above)
Comment (wallyworld) [reply caas/kubernetes/provider/specs/types.go]: ah, right, no double handling naymore

File: caas/kubernetes/provider/k8s_test.go (hunk 2 above)
Comment (ycliuhw) [reply caas/kubernetes/provider/k8s_test.go]: because no need to do 2nd round of yaml unmarshal anymore.

**Comment by ycliuhw** [comment]
$$merge$$
